### ISSUE TITLE
Printer control via touch interface

### ISSUE THREAD

**Issue by xblax** [xblax/flashforge_adm5_klipper_mod#18]
### Discussed in https://github.com/xblax/flashforge_adm5_klipper_mod/discussions/12

Adding this issue as a place-holder for future integration of a touch-screen UI solution using the ADM5 built-in screen. Options discussed are currently:
- Porting [Guppyscreen](https://github.com/ballaswag/guppyscreen). Based on LVGL, so far used as a mod on the Creality K1.
- Porting [CYD-Klipper](https://suchmememanyskill.github.io/CYD-Klipper/). Based on LVGL, so far used with ESP32 and dedicated touch screens.
- Enabling [Klipper Screen](https://klipperscreen.readthedocs.io/en/latest/). Needs X11, libgl, Gtk3 etc.

All discussion around that topic should continue [here](https://github.com/xblax/flashforge_adm5_klipper_mod/discussions/12), until the road map for this feature becomes more clear.

**Comment by consp** [comment]
For future reference: it's now also possible to build a guppyscreen variant (pull feature branch) which is far more lightweight than klipperscreen but lacks some functionality. It's WIP.

**Comment by consp** [comment]
For klipperscreen the changes for the next update (v0.5) will be (merge feature branch):

* Removed error in update (system info) screen #106 #109 
* Added custom fbturbo driver with g2d (see fork consp/xf86-video-fbturbo), since it uses a blitter it speeds the rendering up a bit and should provide the maximum possible experience (since we lack a (E)GL compatible display device).
* Increased niceness of xorg and klipperscreen to make them a bit less nicer than they used to be
* Increased loglevel to error to avoid too excessive logging and hogging memory, this might be the cause of some memory issues.
* Added -OO option to avoid a tiny bit of memory usage
* Wifi and ethernet will now show in the wifi screen #39 

Known issues:
* It's a bit sluggisch, can't be done moch about that without a complete rewrite
* Initial loading of each panel might be a bit slow, subsequent ones will be faster (up to four panels).
* Wifi might not show up due to bugs in the network representation code. If you see no or very few networks: please send us your `/tmp/KlipperScreen.log`! That would be very helpful.

Memory reduction in klipperscreen will be nigh impossible as it will require a rewrite to avoid the massive use of dicts. Good for ease of development, bad for memory they are. To people running large prints I recommend switching to guppyscreen.

For guppyscreen changelog see #134

**Comment by xblax** [comment]
@consp:
Have you already tried if disabling the "UseBackingStore" setting of fbturbo helps with memory usage? https://manpages.opensuse.org/Tumbleweed/xf86-video-fbturbo/fbturbo.4.en.html#Option~4

Last time I looked at memory usage it seemed like the X11 process itself might be holding a copy of the framebuffer.

**Comment by consp** [comment]
> Have you already tried if disabling the "UseBackingStore" setting of fbturbo helps with memory usage? https://manpages.opensuse.org/Tumbleweed/xf86-video-fbturbo/fbturbo.4.en.html#Option~4

The new driver has no backing store at all :) It uses 4-6mb, the original did about 8mb.

**Comment by xblax** [comment]
Yes, refactoring the complete python code does not make much sense.

I've also looked into two "easy" optimizations regarding KlipperScreen, I also pushed them to the buildsystem branch. Maybe you can take a look.

1. I clean up the buildroot packages in the klipperscreen variant a bit, that gives a bit reduction in memory usage (seems to be ~3.5MB less ram usage after boot), it also cuts down the packaged size by almost ~30MB.

3. I wanted to get rid of heavyweight librsvg dependency (that lib is over 7MB). I found this project: https://michael.orlitzky.com/code/libsvgtiny-pixbuf.xhtml I manged to integrate it but was quite some work because I had to integrate the dependencies for that as well. Maybe ~1.5MB RAM saving. But at least I learned more about build root package creation thanks to that :)

I think another way to save a bit more RAM would be to convert the SVGs to PNG in the first place.

**Comment by consp** [comment]
Good 'easy' spots!

> I think another way to save a bit more RAM would be to convert the SVGs to PNG in the first place.

That would definitely be a worthwhile thing to do. SVG's is nice but doesn't make sense on an embedded system.

**Comment by consp** [comment]
> That would definitely be a worthwhile thing to do. SVG's is nice but doesn't make sense on an embedded system.

Just tested it with existing installation (note, also some memory checking module included so don't look at the footprint only the difference).

original svg and librsvg
```
 3981 root     /root/printer_software/Klip      968    37644    39541    43068
```
96dpi png
```
 4062 root     /root/printer_software/Klip        0    34972    37141    40948
```

So some savings, don't see lot of difference visually

**Comment by xblax** [comment]
Klipperscreeen in the mod has been rebased to current Klipperscreen master, SVGs are now replaced with PNGs.

Guppyscreen upstream development seems to have stopped, I don't anticipate that much will be done solve open issues with that variant - I'm using Klipperscreen personally, which seems to run quite stable in the latest release (no print failures due to resource exhaustion).

Open points can be tracked as individual issues, if needed.
